lib/jellyfish/room.ex

```diff
@@ -447,23 +452,115 @@ defmodule Jellyfish.Room do
     {:noreply, state}
   end
 
+  @impl true
+  def handle_info(
+        %EndpointMetadataUpdated{endpoint_id: endpoint_id, endpoint_metadata: metadata},
+        state
+      )
+      when is_map_key(state.peers, endpoint_id) do
+    Logger.info("Peer #{endpoint_id} metadata updated: #{metadata}")
+    Event.broadcast_server_notification({:peer_metadata_updated, state.id, endpoint_id, metadata})
+
+    state = put_in(state, [:peers, endpoint_id, :metadata], metadata)
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_info(%EndpointMetadataUpdated{}, state) do
+    {:noreply, state}
+  end
+
   @impl true
   def handle_info(%TrackAdded{endpoint_id: endpoint_id} = track_info, state)
       when endpoint_exists?(state, endpoint_id) do
-    Logger.info("Endpoint #{endpoint_id} added track #{inspect(track_info)}")
+    endpoint_id_type = get_endpoint_id_type(state, endpoint_id)
+
+    Logger.info("Track #{track_info.track_id} added, #{endpoint_id_type}: #{endpoint_id}")
+
+    Event.broadcast_server_notification(
+      {:track_added, state.id, {endpoint_id_type, endpoint_id}, track_info}
+    )
+
+    endpoint_group = get_endpoint_group(state, track_info.endpoint_id)
+    access_path = [endpoint_group, track_info.endpoint_id, :tracks, track_info.track_id]
+
+    track = Track.from_track_added_message(track_info)
+    state = put_in(state, access_path, track)
 
-    state = add_track(state, track_info)
+    {:noreply, state}
+  end
 
+  @impl true
+  def handle_info(%TrackAdded{endpoint_id: endpoint_id} = track_info, state) do
+    Logger.error("Unknown endpoint #{endpoint_id} added track #{inspect(track_info)}")
+    {:noreply, state}
+  end
```

Comment:
I think this makes sense actually. I think we could expect this message if we delete Peer while it adds new track - we would receive `TrackAdded` and then `EndpointRemoved` messages.

Reply:
If this is the case, then it shouldn't be an error because we expect that this can sometimes happen. I think we should modify the logic in `room.ex`, so we would remove endpoints from the state after we receive `EndpointRemoved.`